[FEAT] 회원가입 요청 API 구현

dye0p opened the pull request:
## Issue Number
closed #211

## As-Is
<!-- 문제 상황 정의 -->
멘토의 회원가입을 위해 회원가입 요청 API가 필요하다.

## To-Be
<!-- 변경 사항 -->
- 회원가입 API를 구현하여 회원 정보를 DB에 저장한다.
- 회원가입시 입력받은 비밀번호를 암호화 하여 DB에 저장한다.
- 회원가입시 입력받은 요소의 유효성을 검증한다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

## (Optional) Additional Description

## Member 테이블 추가
Member 테이블이 추가되었습니다.
<img width="429" height="196" alt="image" src="https://github.com/user-attachments/assets/9376c455-aecb-4693-a219-815a0db874ec" />


## 입력값 검증
- 입력값 검증은 임의로 판단하여 설정하였습니다. 추가 검증 사항이 필요하다면 코멘트 남겨주시면 감사하겠습니다 :)

## 비밀번호 암호화 알고리즘
- 비밀번호 암호화 알고리즘으로 `SHA-256` 알고리즘을 사용하였습니다.

## 비밀번호(Password) 값 객체 사용
처음엔 비밀번호 암호화를 `Member`가 가지도록 하였습니다. 그러나
추후에, 토큰 발급시 추가될 비밀번호 검증과 함께 암호화의 책임이 `Member`에 있는것이 부자연스럽다고 판단했습니다.
해당 근거로 `Password` 라는 값 객체를 생성하여 책임을 분리하였습니다.

값 타입을 컬럼(**VARCHAR**)으로 적용하기 위해서 `@Converter + 불변 객체` 조합을 사용하였습니다.

**+ 8/1 추가 수정**
- @Converter + 불변 객체 사용시 암호화 되지않은 비밀번호 객체를 만들 수 있는 이슈가 있어서 `@Embeddable + @Embeddable` 조합을 사용하여 값 타입을 다루도록 변경하였습니다.

## 참고
`@Embeddable`, `@Converter` 비교
- 두 어노테이션 모두 JPA에서 제공하는 어노테이션 입니다.
- `@Embeddable` 은 객체를 값 타입으로 사용할 수 있게 해주지만, 객체의 필드를 불변으로 선언하지 못합니다.
- 값 객체는 객체의 불변을 보장해줘야하기 때문에 필드를 불변으로 선언할 수 있는 `@Converter + 불변객체` 조합을 사용하였습니다.

dye0p committed:
feat: 회원가입 API 추가

dye0p committed:
feat: 회원가입 입력 유효성 검사 추가

dye0p committed:
feat: 멤버 저장 로직 트랜잭션 추가

dye0p committed:
feat: 비밀번호 암호화 기능 추가

- 기존 회원가입시 필요한 비밀번호가 평문으로 저장되어, 보안상의 문제가 있었다. 이를 보완하기 위해 DB에 저장하기 전, 비밀번호를 암호화 하는 로직을 추가하였다.

dye0p commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/password/PasswordConverter.java:
```diff
@@ -0,0 +1,18 @@
+import jakarta.persistence.AttributeConverter;
+import jakarta.persistence.Converter;
+
+@Converter
+public class PasswordConverter implements AttributeConverter<Password, String> {
+
+    @Override
+    public String convertToDatabaseColumn(Password attribute) {
+        return attribute.getPassword();
+    }
+
+    @Override
+    public Password convertToEntityAttribute(String dbData) {
+        return Password.formEncrypt(dbData);
+    }
+}
```
해당 클래스는 `AttributeConverter`를 구현하는 구현체로 

`convertToDatabaseColumn()`
- 객체(Password) -> db 컬럼(String)

`convertToEntityAttribute()`
- db 컬럼(String) -> 객체(Password)

로 변환 해주는 기능을 수행합니다.

dye0p commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/password/Password.java:
```diff
@@ -0,0 +1,43 @@
+
+    private static String encrypt(String password) {
+        try {
+            final String algorithm = "SHA-256";
```
`final` 키워드를 명시해준 이유는, 컨벤션과는 별개로 해당 지역변수가 **암호화에 필요한 알고리즘을 가지는 변수** 이고, 해당 로직 안에서 사용될 때 **변경되면 안된다는 것**을 명시적으로 나타내주기 위해서 사용하였습니다!

납득이 안되면 말씀해주세요..

kimdevv replied:
p3: 알고리즘을 상수 처리하는 것과는 차이가 있을까요? `SHA-256` 알고리즘을 선택하신 이유도 궁금합니다!

dye0p replied:
처음엔 알고리즘을 상수로 선언하였지만, 현재 하나의 메서드에서만 사용된다는 점에서 굳이 상수화를 해야할까 싶었습니다. 
메서드 내부에서 지역변수로 선언하여도 의미를 잘 전달할 수 있다고 판단했습니다!

> SHA-256 알고리즘을 선택하신 이유도 궁금합니다!

특별한 이유는 없었습니다. 알고있던 암호화 알고리즘이 SHA-256 알고리즘 뿐이었습니다.

dye0p committed:
feat: 아이디 중복검증 API 구현

- 아이디 중복검증이 필요하기 때문에 API를 구현하였음.
- Member 엔티티의 loginId 컬럼이 unique 제약조건 때문에 중복 저장이 되지 않기 때문에, register 메서드 호출시에도 중복검증을 해주도록 하였음

jye326 commented on backend/fittoring/src/main/java/fittoring/exception/GlobalExceptionHandler.java:
```diff
@@ -25,6 +25,11 @@ public ResponseEntity<ErrorResponse> handle(CategoryNotFoundException e) {
     }
 
+    @ExceptionHandler(DuplicateIdException.class)
+    public ResponseEntity<ErrorResponse> handleDuplicateId(DuplicateIdException e) {
```
p3 : 굳이 디테일한 메서드 이름을 지어줘야 할까요?저는 어차피 다 같은 handler고 처리하는 exception 으로 구별이 가능해서 handle로 통일했습니다.

dye0p replied:
그렇네요! 통일하겠습니다.

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/password/Password.java:
```diff
@@ -0,0 +1,43 @@
+
+    public static Password createEncrypt(String password) {
+        return new Password(encrypt(password));
+    }
```
p2 : 
1. 왜 암호화 안 된 그냥 문자열을 갖고 있는 Password 객체를 생성해야 한가요?  암호화 되지 않은 문자열 password 를 갖고 있는 객체가 존재하지 않도록 fromEncrypt는 삭제해야 될 것 같아요.
2. createEncypt : '암호화를 생성한다' 보다는 패스워드는 암호화를 전제로 하고 "createPassword" 로 수정하는 것은 어떨까요? 반환타입도 Password 니까요!

dye0p replied:
동의합니다! 1번의 이유가 고민 포인트인 것 같아요. 해당 메서드가 생긴 이유는 암호화된 String 타입의 문자열을 그대로 Passwrod 객체로 만들어주기 위함이었습니다. 기존에 Embeddable 타입의 객체를 만들지 않았던 이유는 값 객체의 상태를 불변으로 보장해주지 못하는 단점 때문에 사용하지 않았는데, final 필드로 선언하고, 기본생성자를 null로 초기화 해준다면 불변 필드로 사용가능 할 것 같습니다. 

정리하자면
1. 기존 방식인 @Converter + 불변 객체 사용방식을 버린다.
2. @Embeddable + final 필드를 사용하여 값 객체로 사용한다.

ppparkta replied:
final 붙여서 만들 수 있나요? 저도 궁금해요

dye0p replied:
``` java
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Embeddable
public class Password {

    private final String password;

    protected Password() {
        this.password = null;
    }

```
필드를 final로 선언하면 무조건 객체의 생성시점에 초기화 해줘야 합니다.
JPA는 객체를 생성할 때 기본생성자를 이용하는데, 여기서 컴파일 에러가 안나기 위해서 우선은 null로 초기화 되도록 합니다.
이렇게 구성하면 코드 상으로 불변처럼 사용할 수 있을 것 같습니다.

추가로, 제가 리플렉션에 대해서 정확하게 알지 못해서 알아봤습니다.

JPA에서 기본 생성자를 이용하여 먼저 상태가 null인 객체를 생성하고, 리플렉션을 이용해서 필드에 값을 주입한다고 합니다. 

> 그런데 필드가 final 이라서 주입을 못하지 않나?

라는 의문이 들었는데, 리플렉션에서 final 필드에 값을 재할당 하려면 별도의 설정이 필요한데, 

Hibernate 6에서는 Java의 버전에 따라서 **리플렉션 우회**를 명시적으로 허용한다고 합니다.
그래서 final 필드에 재할당이 가능하다고 합니다.
(그런데 JPA에서 이렇게 사용하는건 권장하지 않는다고 하네요..)

일단은, 지금처럼 사용은 가능할 것 같습니다.

처음 알게된 부분이었는데, 참고자료 포함하겠습니다!
[Change private static final field using Java reflection](https://stackoverflow.com/questions/3301635/change-private-static-final-field-using-java-reflection)

jye326 replied:
수고하셨습니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/password/Password.java:
```diff
@@ -0,0 +1,43 @@
+            return hex.toString();
+
+        } catch (NoSuchAlgorithmException e) {
+            throw new IllegalArgumentException("비밀번호 암호화에 실패했습니다.");
```
p2 : 커스텀 예외 처리 해주세요~

dye0p replied:
네 확인했습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Member.java:
```diff
@@ -0,0 +1,51 @@
+    private String loginId;
+
+    @Column(nullable = false)
+    private String male;
```
p2: 성별을 나타내는 칼럼이라면 이름을 `gender`으로 하는 건 어떨까요?

dye0p replied:
gender가 더욱 직관적이네요! 변경하겠습니다

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Member.java:
```diff
@@ -0,0 +1,51 @@
+    @Column(nullable = false)
+    private String name;
+
+    @Column(nullable = false)
```
p2: 동일한 폰 번호로는 가입하지 못하도록 `unique`인 게 좋을 것 같습니다.

dye0p replied:
좋네요! 추가하겠습니다.

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Member.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Convert(converter = PasswordConverter.class)
+    @Column(nullable = false)
+    private Password password;
```
p3: 비밀번호를 포장하신 이유가 궁금합니다! DB에 I/O가 발생할 때마다 Convert하게 될 텐데, 그냥 `String`으로 관리해도 괜찮지 않을까 생각했습니다!

dye0p replied:
비밀번호를 포장한 이유는 다음과 같습니다.
- 비밀번호 암호화와 같은 관리의 책임이 Member에게 있는 것이 부자연스럽다고 판단했습니다.
- 메서드에서 비밀번호를 매개변수로 받아야한다면, String 보단, Password 타입으로 받는것이 안전하다 판단하였습니다.

비밀번호를 포장하지 않고, String 으로 관리한다고 하면 암호화 같은 로직은 유틸성 클래스로 분리해볼 수 있을 것 같습니다. 

코멘트에 대한 궁금한 부분이 있어 질문 드립니다.
DB에 I/O가 발생할때 마다 Convert하게 되면 우려되는 부분이 있나요?

kimdevv replied:
매번 Convert 하게 되면 시간이 조금 더 걸리지 않을까 생각하긴 했는데, 유의미할 정도로 차이가 크진 않을 것 같습니다!

결국 DB에는 암호화된 비밀번호의 문자열이 저장될 텐데 왜 Password로 포장을 하신 건지 의도가 궁금했습니다! 매개변수를 모든 `String`으로 받는 게 아니라 Password라는 타입으로 한 번 거를 수 있다는 점에서 장점이라고 생각되네요~

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Member.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public static Member of(String loginId, String male, String name, String phone, String password) {
+        return new Member(loginId, male, name, phone, Password.createEncrypt(password));
```
p3: 비밀번호를 암호화하라는 명령을 `Member`가 하고 있는 상황인데, 암호화하는 책임을 바깥으로 빼내서 암호화된 비밀번호를 받는 정적 팩토리 메서드를 만드는 건 어떤가요??

dye0p replied:
좋은 의견인 것 같습니다. 한가지 고민되는 부분이 있는데, Password를 바깥에서 받기위해서 외부에서 암호화를 해주어야 하는데, 어디서 해주는 것이 적절하다 생각하시나요?

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/AuthService.java:
```diff
@@ -0,0 +1,30 @@
+        memberRepository.save(member);
+    }
+
+    @Transactional(readOnly = true)
```
p3: 조회 쿼리 단 하나만 날리는 메서드라서 트랜잭션으로 묶는 것이 아무런 효과가 없을 것 같은데, 혹시 어떻게 생각하시나요?!

dye0p replied:
죄송합니다. 질문의 의도를 잘 이해하지 못했습니다.. 말씀해주신 의도는 **트랜잭션으로 묶지 않아도 될 것 같다.**라는 의도로 말씀하신 걸까요?

생각해보니, 해당 메서드는 단순 조회만 하고 있기에 트랜잭션이 필요없어 보입니다! 

혹시 조회 로직에서 트랜잭션으로 묶는 기준이 있으신가요?

kimdevv replied:
저 같은 경우는 한 메서드에 쿼리문이 여러 개 있는 경우에는 팬텀 리드와 같은 부정합을 막기 위해서 트랜잭션으로 묶는데요! 조회 쿼리뿐만 아니라 쿼리문 하나만 메서드에서 사용되는 경우에는 그 메서드를 트랜잭션으로 묶던 안 묶던 아무 차이가 없다고 생각했습니다!

dye0p replied:
좋은 의견 감사합니다 : )

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/dto/SignUpRequest.java:
```diff
@@ -0,0 +1,33 @@
+
+public record SignUpRequest(
+
+        @Size(min = 4, max = 20, message = "아이디는 5자 이상 15자 이하로 입력해주세요.")
```
p1: `min`과 `max`가 `message`의 내용과 다른 것 같습니다!

dye0p replied:
감사합니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -0,0 +1,31 @@
+        return new ResponseEntity<>(HttpStatus.CREATED);
+    }
+
+    @PostMapping("/validate-id")
+    public ResponseEntity<Void> validateDuplicateId(@RequestBody @Valid ValidateDuplicateIdRequest request) {
+        authService.validateDuplicateId(request.id());
```
p3: `id`라는 이름이 `Member` 엔티티의 pk값인지, 아니면 loginId인지 조금 헷갈렸습니다! `ValidateDuplicateIdRequest` dto에서도 마찬가지구요! 엔티티에서 사용하고 있는 네이밍인 `loginId`를 사용하는 건 어떨까요??

dye0p replied:
맞습니다. PK와 느낌이 비슷하여 헷갈릴것 같습니다. 어떤 네이밍을 할지 고민이 되는데,

회원가입에서 입력받는 아이디는 회원가입 아이디(signUpId)의 의미로 사용될 수 있습니다.
하지만, Member의 필드는 loginId입니다. 

signUpId와 loginId중 어느것이 더 적절하다 생각하시나요?

jye326 replied:
signUp이 성공하면 그 아이디가 loginId로 저장될테니까 loginId로 통일해도 괜찮지 않을까요?

kimdevv replied:
저도 하나로 통일할 거면 `loginId`가 좋을 것 같습니다! 아니면 애초에 이런 혼동이 생기지 않도록 `username` 같은 네이밍은 어떤가요??

dye0p replied:
> username 같은 네이밍은 어떤가요??

username에 대해서 생각해보면 name 필드와 혼란이 올 것이 우려됩니다!..

그래서 longId로 통일하겠습니다!

좋은 의견 감사합니다 : )

kimdevv reviewed (commented):
고생하셨습니다!!

dye0p committed:
refactor: Member 성별 필드 네이밍 변경(male -> gender)

dye0p committed:
fix: 회원가입 요청 dto 아이디 제약사항 정확하게 수정

dye0p committed:
feat: Member phone 필드 unique 제약사항 추가

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Member.java:
```diff
@@ -0,0 +1,51 @@
+public class Member {
+
+    @Id
+    @GeneratedValue(strategy = jakarta.persistence.GenerationType.IDENTITY)
```
p2 : 얘도 다른 엔티티들 처럼 jakarta.persistence.GenerationType import 해도 될 것 같아요.

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/dto/SignUpRequest.java:
```diff
@@ -0,0 +1,33 @@
+        String loginId,
+
+        @NotBlank
+        @Size(min = 2, max = 5, message = "이름은 2자 이상 5자 이하로 입력해주세요.")
```
p3 : name만 어노테이션 순서가 다르네요. 맞춰볼까요?

dye0p replied:
꼼꼼하시네요 :)

jye326 reviewed (commented):
코멘트 확인 부탁드립니다~

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/password/Password.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final String password;
+
+    public static Password formEncrypt(String password) {
```
p3: 단순 생성자 역할을 하는 것 같은데, `from`과 같은 관습적인 네이밍은 어떨까요?

dye0p replied:
동의합니다! 변경하겠습니다.

kimdevv reviewed (approved):
수고하셨습니다~

dye0p committed:
refactor: 어노테이션 지정 순서 변경

- 어노테이션 지정 순서 일관성을 위해 변경

dye0p committed:
refactor: ID(PK) 생성전략 클래스 import

dye0p committed:
refactor: Member 생성방식 변경

- Member의 생성자 내부에서 비밀번호 암호화 로직을 수행했지만 외부에서 주입해줄 때 Password 타입을 강제할 수 없다는 단점이 있었음
이러한 이유로 기존 방식을, 암호화된 비밀번호를 외부에서 주입받도록 변경하였음

dye0p committed:
refactor: Password 객체 Embeddable 타입으로 변경

- 기존 로직에서 convertToEntityAttribute() 로직에서 비밀번호를 그대로 받아서 Password를 사용하는 로직을 사용했음. 여기서 외부의 String을 받아서 그대로 객체를 생성하는 정적 팩터리 메서드가 열려있다는 것이 우려되었음
- 기존의 Converter 방식에서 Embeddable 타입으로 변경하였음

dye0p committed:
refactor: 로그인 ID 검증과 관련된 객체와 변수들의 네이밍 명확하게 변경

- 기존 id로 표현되어서 PK의 id와 혼란될 수 있다고 판단하여 명확하게 변경

dye0p committed:
feat: 비밀번호 암호화 실패 커스텀 예외 추가

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/password/Password.java:
```diff
@@ -0,0 +1,46 @@
+
+            StringBuilder hex = new StringBuilder();
+
+            for (byte b : digest) {
+                hex.append(String.format("%02x", b));
+            }
+            return hex.toString();
```
p3: 이  코드는 SmsClientService의 시그니처 생성을 위한 16진수 변환코드와 완벽히 일치하는 것 같아요.
유틸로 빼는거 어떠신가요?

dye0p replied:
꼼꼼하게 확인하지 못했네요..! 동일하게 맞다면 공통으로 빼도 좋을 것 같습니다. 확인 후 반영하겠습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/AuthService.java:
```diff
@@ -0,0 +1,40 @@
+
+    public void validateDuplicateLoginId(String loginId) {
+        if (memberRepository.existsByLoginId(loginId)) {
+            throw new DuplicateLoginIdException("이미 사용 중인 아이디입니다. 다른 아이디를 입력해주세요.");
```
p2: 메시지 enum에 추가해주세요!

dye0p replied:
꼼꼼하시네요 : )

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/dto/SignUpRequest.java:
```diff
@@ -0,0 +1,28 @@
+
+public record SignUpRequest(
+
+        @Size(min = 5, max = 15, message = "아이디는 5자 이상 15자 이하로 입력해주세요.")
+        @NotBlank
+        String loginId,
```
p2: 매개변수에 어노테이션이 붙을 경우 상단이 아니라 좌측 배치하기로 논의했던걸로 기억해요! 
길어서 올라간건 이해되는데 매개변수에 공백 추가도 그렇고 상의해봐야 할거같습니다~~

dye0p replied:
맞습니다. 컨벤션 인지는 했지만, 너무 길어서 위로 올릴 수 밖에 없었어요..

ppparkta commented:
꼼꼼한 테스트 좋네요. 궁금한 부분 코멘트 남겼어요. 고생하셨습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/password/Password.java:
```diff
@@ -0,0 +1,46 @@
+            return hex.toString();
+
+        } catch (NoSuchAlgorithmException e) {
+            throw new PasswordEncryptionException("비밀번호 암호화에 실패했습니다.");
```
p2: 여기도 enum에 빼주시면 좋을 것 같습니다

dye0p replied:
이 부분도 놓쳤네요..!

dye0p committed:
refactor: 예외 메세지 상수화

dye0p reviewed (commented):
다들 꼼꼼하게 리뷰 해주셨네요 감사합니다 :)

jye326 commented on backend/fittoring/src/test/java/fittoring/integration/mentoring/api/AuthControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+                .post("/signup");
+
+        //then
+        assertThat(response.statusCode()).isEqualTo(400);
```
p3: 어차피 RestAssured를 쓰고 있으니까 assertThat()으로 잡기보다 .then()으로 체이닝하여 상태코드와 메시지를 동시에 확인하는 것은 어떨까요?
```
    .then()
    .statusCode(400)
    .body("message", containsString("username is required"));
```
이런 느낌으로다가

jye326 reviewed (approved):
깊은 고민이 느껴지네요. 수고하셨습니다.

dye0p committed:
refactor: 중복된 헥스 변환 로직을 HexEncoder 클래스로 추출

- 헥스 변환 로직이 중복 사용되고 있었기 때문에 하나의 유틸 클래스로 추출하여 재사용성을 개선하였음

dye0p committed:
test: HexEncoder 테스트 추가

dye0p committed:
refactor: 불필요한 공백 제거 (팀 컨벤션)
